src/node_contextify.cc

```diff
@@ -1816,6 +1818,18 @@ bool ShouldRetryAsESM(Realm* realm,
   Utf8Value message_value(isolate, message);
   auto message_view = message_value.ToStringView();
 
+  for (const auto& error_message : throws_only_in_cjs_error_messages) {
+    if (message_view.find(error_message) != std::string_view::npos) {
+      isolate->ThrowException(v8::Exception::SyntaxError(
+          String::NewFromUtf8(
+              isolate,
+              "Top-level await is not supported in CommonJS. "
+              "Consider using ESM or wrap await in an async function.")
+              .ToLocalChecked()));
```

Comment:
It seems like this message is being printed for _any_ of the “throws only in CJS” error messages? I would think that we would want to print this only when the user wrote a top-level `await` in a file that didn’t parse successfully as ESM.

Also in general our messages don’t use “ESM” as that’s not a widely known term. I think we usually write “module syntax” or “import/export syntax”.

Reply:
thank you very much for the improvement suggestions, I wonder if you could comment on the broken test/es-module/test-esm-detect-ambiguous.mjs tests, I am trying to improve them

in addition, I wonder if it would be better to include my newly added test/parallel/test-commonjs-top-level-await.js test in test-esm-detect-ambiguous.mjs

Reply:
What test failure do you see?

> in addition, I wonder if it would be better to include my newly added test/parallel/test-commonjs-top-level-await.js test in test-esm-detect-ambiguous.mjs

Yes, my preference would be to keep all the syntax detection tests together. If you could follow the pattern of the other related tests that would be great.

Reply:
I saw this  
https://github.com/nodejs/node/actions/runs/11863231944/job/33064268346?pr=55874

Reply:
> I saw this
> [nodejs/node/actions/runs/11863231944/job/33064268346?pr=55874](https://github.com/nodejs/node/actions/runs/11863231944/job/33064268346?pr=55874)

I don't see the failure in this output. When you run the tests locally, what fails?

Reply:
@GeoffreyBooth would you have any suggestions for these test failures?

Reply:
I think those failures are probably legit and the implementation needs to be adjusted so that they pass.

Reply:
> I think those failures are probably legit and the implementation needs to be adjusted so that they pass.

thank you, as you said I will try to fix the tests as legitimate

Reply:
I made an update for this place, many thanks for the suggestions @GeoffreyBooth